### PR TITLE
Validate bintray credentials on demand

### PR DESCRIPTION
**Pull request by markusjura**
So far have checked for valid Bintray credentials every time the sandbox has been started or a bundle has been loaded from Bintray.

Due to the fact that the ConductR artifacts are available on a public Bintray repository, we don’t need to perform this check every time. Also, every bundle is hosted on a public Bintray repository, including the monitoring bundles (https://bintray.com/lightbend/commercial-monitoring is also public). As a result, the user doesn’t need to setup Bintray credentials anymore from onwards ConductR 2.1.0.

This PR is:
- Adding a new validation to `sandbox_run_jvm` that checks based on the ConductR version if Bintray credentials are needed. If Bintray credentials are needed and the credential file is not created or malformed then the `sandbox run` command will abort before starting ConductR
- Changing the Bintray resolver so that no exception is thrown anymore if the credentials could not be loaded. Instead, we now try to download the artifacts without credentials. If a `404 Not Found` with the response text `Repo my-repo was not found` then we will catch this error and abort with the message:

    ```
    $ conduct load lightbend/commercial-releases/ConductR-Agent-Debian
    Retrieving bundle..
    Loading bundle from cache lightbend/commercial-releases/ConductR-Agent-Debian
    Error: Unable to find Bintray repository lightbend/commercial-releases. If this is a private repository make sure to setup the Bintray credentials at /Users/mj/.lightbend/commercial.credentials
    ```

Note that this message is only thrown for private Bintray repositories and after our `sandbox run` validation check. So for all ConductR artificats hosted on Bintray this error should never occur. The error should only occur in case the user wants to load bundles from a custom private repository.

### CODE REVIEW

**Comment by markusjura** [comment]
**Checklist**
- [x] Implementation
- [x] Manual tests
- [x] Unit tests

**Comment by markusjura** [comment]
## Manual tests

### ConductR 2.1.0 - No Bintray credentials (loading from cache because version not yet published to Bintray)

```
$ sandbox run 2.1.0 --feature monitoring
|------------------------------------------------|
| Starting ConductR                              |
|------------------------------------------------|
Extracting ConductR core to /Users/mj/.conductr/images/core
Extracting ConductR agent to /Users/mj/.conductr/images/agent
Starting ConductR core instance on 192.168.10.1..
Waiting for ConductR to start..
Starting ConductR agent instance on 192.168.10.1..
|------------------------------------------------|
| Starting HAProxy                               |
|------------------------------------------------|
Exposing the following ports [80, 443, 3000, 5601, 8999, 9000, 9200, 9999]
87c3db7ce7971d55b7f475021defc9588aeac5ea1cb6320366ead2cadc7a1946
Deploying bundle conductr-haproxy with configuration conductr-haproxy-dev-mode
Retrieving bundle..
Loading bundle from cache typesafe/bundle/conductr-haproxy
Retrieving file:///Users/mj/workspace/conductr-cli/conductr-haproxy
Resolving bundle typesafe/bundle/conductr-haproxy
Retrieving https://dl.bintray.com/typesafe/bundle/typesafe/conductr-haproxy/v2-bdfa43d7ade7456af30a95e9ef4bce863d3029dc58b6824f5783881297d1983b/conductr-haproxy-v2-bdfa43d7ade7456af30a95e9ef4bce863d3029dc58b6824f5783881297d1983b.zip
[#################################################] 100%
Retrieving configuration..
Loading bundle configuration from cache typesafe/bundle-configuration/conductr-haproxy-dev-mode
Retrieving file:///Users/mj/workspace/conductr-cli/conductr-haproxy-dev-mode
Resolving bundle configuration typesafe/bundle-configuration/conductr-haproxy-dev-mode
Retrieving https://dl.bintray.com/typesafe/bundle-configuration/typesafe/conductr-haproxy-dev-mode/v2-e5f35046d59c27f4cfeebcc026f3e3b76e027b15223a8961b59d728e559402e2/conductr-haproxy-dev-mode-v2-e5f35046d59c27f4cfeebcc026f3e3b76e027b15223a8961b59d728e559402e2.zip
[#################################################] 100%
Loading bundle to ConductR..
[#################################################] 100%
Bundle bdfa43d7ade7456af30a95e9ef4bce86-e5f35046d59c27f4cfeebcc026f3e3b7 is installed
Bundle loaded.
Bundle run request sent.
Bundle bdfa43d7ade7456af30a95e9ef4bce86-e5f35046d59c27f4cfeebcc026f3e3b7 waiting to reach expected scale 1
Bundle bdfa43d7ade7456af30a95e9ef4bce86-e5f35046d59c27f4cfeebcc026f3e3b7 has scale 0, expected 1.
Bundle bdfa43d7ade7456af30a95e9ef4bce86-e5f35046d59c27f4cfeebcc026f3e3b7 expected scale 1 is met
|------------------------------------------------|
| Starting logging feature based on              |
| elasticsearch and kibana                       |
|------------------------------------------------|
conductr-kibana bundle is packaged as a Docker image. Checking Docker requirements..
Docker is installed and configured correctly.
Deploying bundle conductr-elasticsearch..
Retrieving bundle..
Loading bundle from cache typesafe/bundle/conductr-elasticsearch
Retrieving file:///Users/mj/workspace/conductr-cli/conductr-elasticsearch
Resolving bundle typesafe/bundle/conductr-elasticsearch
Retrieving https://dl.bintray.com/typesafe/bundle/typesafe/conductr-elasticsearch/v1-85dd2657bf86e5ed817c6cbe9d4c18e3a6c30eb3146383fc65357b65f11cb550/conductr-elasticsearch-v1-85dd2657bf86e5ed817c6cbe9d4c18e3a6c30eb3146383fc65357b65f11cb550.zip
[#################################################] 100%
Loading bundle to ConductR..
[#################################################] 100%
Bundle 85dd2657bf86e5ed817c6cbe9d4c18e3 is installed
Bundle loaded.
Bundle run request sent.
Bundle 85dd2657bf86e5ed817c6cbe9d4c18e3 waiting to reach expected scale 1
Bundle 85dd2657bf86e5ed817c6cbe9d4c18e3 has scale 0, expected 1..
Bundle 85dd2657bf86e5ed817c6cbe9d4c18e3 expected scale 1 is met
Deploying bundle conductr-kibana..
Retrieving bundle..
Loading bundle from cache typesafe/bundle/conductr-kibana
Retrieving file:///Users/mj/workspace/conductr-cli/conductr-kibana
Resolving bundle typesafe/bundle/conductr-kibana
Retrieving https://dl.bintray.com/typesafe/bundle/typesafe/conductr-kibana/v1-06d370bcf909477a661e19a2b3fef3d3d9955cfdeca5d3f8991945d8f0904478/conductr-kibana-v1-06d370bcf909477a661e19a2b3fef3d3d9955cfdeca5d3f8991945d8f0904478.zip
[#################################################] 100%
Loading bundle to ConductR..
[#################################################] 100%
Bundle 06d370bcf909477a661e19a2b3fef3d3 is installed
Bundle loaded.
Bundle run request sent.
Bundle 06d370bcf909477a661e19a2b3fef3d3 waiting to reach expected scale 1
Bundle 06d370bcf909477a661e19a2b3fef3d3 has scale 0, expected 1..
Bundle 06d370bcf909477a661e19a2b3fef3d3 expected scale 1 is met
|------------------------------------------------|
| Starting monitoring feature                    |
|------------------------------------------------|
Deploying bundle cinnamon-grafana-docker..
Retrieving bundle..
Loading bundle from cache typesafe/bundle/cinnamon-grafana-docker
Retrieving file:///Users/mj/workspace/conductr-cli/cinnamon-grafana-docker
Resolving bundle typesafe/bundle/cinnamon-grafana-docker
Retrieving https://dl.bintray.com/typesafe/bundle/typesafe/cinnamon-grafana-docker/v2-cfa405be2b6f46b5608b8eb5d38d18861a0b1cbc317206d39f369027914b7c7c/cinnamon-grafana-docker-v2-cfa405be2b6f46b5608b8eb5d38d18861a0b1cbc317206d39f369027914b7c7c.zip
[#################################################] 100%
Loading bundle to ConductR..
[#################################################] 100%
Bundle cfa405be2b6f46b5608b8eb5d38d1886 is installed
Bundle loaded.
Bundle run request sent.
Bundle cfa405be2b6f46b5608b8eb5d38d1886 waiting to reach expected scale 1
Bundle cfa405be2b6f46b5608b8eb5d38d1886 has scale 0, expected 1..
Bundle cfa405be2b6f46b5608b8eb5d38d1886 expected scale 1 is met
|------------------------------------------------|
| Summary                                        |
|------------------------------------------------|
|- - - - - - - - - - - - - - - - - - - - - - - - |
| ConductR                                       |
|- - - - - - - - - - - - - - - - - - - - - - - - |
ConductR has been started:
  core instance on 192.168.10.1
  agent instance on 192.168.10.1
ConductR service locator has been started on:
  192.168.10.1:9008
|- - - - - - - - - - - - - - - - - - - - - - - - |
| Proxy                                          |
|- - - - - - - - - - - - - - - - - - - - - - - - |
HAProxy has been started
By default, your bundles are accessible on:
  192.168.10.1:9000
|- - - - - - - - - - - - - - - - - - - - - - - - |
| Features                                       |
|- - - - - - - - - - - - - - - - - - - - - - - - |
The following feature related bundles have been started:
  conductr-kibana on 192.168.10.1:5601
  conductr-elasticsearch on 192.168.10.1:9200
  cinnamon-grafana-docker on 192.168.10.1:3000
|- - - - - - - - - - - - - - - - - - - - - - - - |
| Bundles                                        |
|- - - - - - - - - - - - - - - - - - - - - - - - |
Check latest bundle status with:
  conduct info
Current bundle status:
Licensed To: cc64df31-ec6b-4e08-bb6b-3216721a56b@lightbend
Max ConductR agents: 3
ConductR Version(s): 0.1.0, 2.1.*
Grants: akka-sbr, cinnamon, conductr

ID               NAME                     VER  #REP  #STR  #RUN  ROLES
bdfa43d-e5f3504  conductr-haproxy          v2     1     0     1  haproxy
85dd265          conductr-elasticsearch    v1     1     0     1  elasticsearch
06d370b          conductr-kibana           v1     1     0     1  kibana
cfa405b          cinnamon-grafana-docker   v2     1     0     1  web
```

### No Bintray credentials - version not found on Bintray

```
$ sandbox run 2.2.0 --feature monitoring
|------------------------------------------------|
| Starting ConductR                              |
|------------------------------------------------|
Error: ConductR core artifact 2.2.0 cannot be found on Bintray.
Error: Please specify a valid ConductR version.
Error: The latest version can be found on: https://www.lightbend.com/product/conductr/developer
```

### ConductR 2.0.5 - No Bintray credentials

```
$ sandbox run 2.0.5 --feature monitoring
Error: Nearly there! The ConductR artifacts are hosted on private Bintray repository.
Error: It is therefore necessary to create a Bintray credentials file at /Users/mj/.lightbend/commercial.credentials
Error: For more information how to setup the Lightbend Bintray credentials please follow:
Error:   http://developers.lightbend.com/docs/reactive-platform/2.0/setup/setup-sbt.html
```

### ConductR 2.0.5 - No Bintray credentials - offline mode

```
$ export CONDUCTR_OFFLINE_MODE=true
$ sandbox run 2.0.5
|------------------------------------------------|
| Starting ConductR                              |
|------------------------------------------------|
Extracting ConductR core to /Users/mj/.conductr/images/core
Extracting ConductR agent to /Users/mj/.conductr/images/agent
Starting ConductR core instance on 192.168.10.1..
Waiting for ConductR to start..
Starting ConductR agent instance on 192.168.10.1..
|------------------------------------------------|
| Starting HAProxy                               |
|------------------------------------------------|
Exposing the following ports [80, 443, 3000, 5601, 8999, 9000, 9200, 9999]
d871fc27f2f14d62d5b1661cf171f20d501296c2116e1966211377cf85c6aa7b
Deploying bundle conductr-haproxy with configuration conductr-haproxy-dev-mode
Retrieving bundle..
Retrieving from cache /Users/mj/.conductr/cache/bundle/conductr-haproxy-v2-bdfa43d7ade7456af30a95e9ef4bce863d3029dc58b6824f5783881297d1983b.zip
Retrieving configuration..
Retrieving from cache /Users/mj/.conductr/cache/configuration/conductr-haproxy-dev-mode-v2-e5f35046d59c27f4cfeebcc026f3e3b76e027b15223a8961b59d728e559402e2.zip
Loading bundle to ConductR..
[#################################################] 100%
Bundle bdfa43d7ade7456af30a95e9ef4bce86-e5f35046d59c27f4cfeebcc026f3e3b7 is installed
Bundle loaded.
Bundle run request sent.
Bundle bdfa43d7ade7456af30a95e9ef4bce86-e5f35046d59c27f4cfeebcc026f3e3b7 waiting to reach expected scale 1
Bundle bdfa43d7ade7456af30a95e9ef4bce86-e5f35046d59c27f4cfeebcc026f3e3b7 has scale 0, expected 1...
Bundle bdfa43d7ade7456af30a95e9ef4bce86-e5f35046d59c27f4cfeebcc026f3e3b7 expected scale 1 is met
|------------------------------------------------|
| Starting logging feature based on eslite       |
|------------------------------------------------|
Deploying bundle eslite..
Retrieving bundle..
Retrieving from cache /Users/mj/.conductr/cache/bundle/eslite-v1-3349b6bfbc09e8d88123fee5b33b2717e07fb2a91de719c99e0403c9e4b28ea5.zip
Loading bundle to ConductR..
[#################################################] 100%
Bundle 3349b6bfbc09e8d88123fee5b33b2717 is installed
Bundle loaded.
Bundle run request sent.
Bundle 3349b6bfbc09e8d88123fee5b33b2717 waiting to reach expected scale 1
Bundle 3349b6bfbc09e8d88123fee5b33b2717 has scale 0, expected 1.
Bundle 3349b6bfbc09e8d88123fee5b33b2717 expected scale 1 is met
|------------------------------------------------|
| Summary                                        |
|------------------------------------------------|
|- - - - - - - - - - - - - - - - - - - - - - - - |
| ConductR                                       |
|- - - - - - - - - - - - - - - - - - - - - - - - |
ConductR has been started:
  core instance on 192.168.10.1
  agent instance on 192.168.10.1
ConductR service locator has been started on:
  192.168.10.1:9008
|- - - - - - - - - - - - - - - - - - - - - - - - |
| Proxy                                          |
|- - - - - - - - - - - - - - - - - - - - - - - - |
HAProxy has been started
By default, your bundles are accessible on:
  192.168.10.1:9000
|- - - - - - - - - - - - - - - - - - - - - - - - |
| Bundles                                        |
|- - - - - - - - - - - - - - - - - - - - - - - - |
Check latest bundle status with:
  conduct info
Current bundle status:
ID               NAME              VER  #REP  #STR  #RUN  ROLES
bdfa43d-e5f3504  conductr-haproxy   v2     1     0     1  haproxy
3349b6b          eslite             v1     1     0     1  elasticsearch
```

### conduct load - public repository - No Bintray credentials

```
$ conduct load visualizer
Retrieving bundle..
Loading bundle from cache typesafe/bundle/visualizer
Bintray credentials loaded from /Users/mj/.lightbend/commercial.credentials
Retrieving file:///Users/mj/workspace/conductr-cli/visualizer
Resolving bundle typesafe/bundle/visualizer
Bintray credentials loaded from /Users/mj/.lightbend/commercial.credentials
Retrieving https://dl.bintray.com/typesafe/bundle/typesafe/visualizer/v2-73595ecbdad1f01a05db5304046b4ad5c0a98ef83e201e507a153cea21087078/visualizer-v2-73595ecbdad1f01a05db5304046b4ad5c0a98ef83e201e507a153cea21087078.zip
[#################################################] 100%
Loading bundle to ConductR..
[#################################################] 100%
Bundle 73595ecbdad1f01a05db5304046b4ad5 is installed
Bundle loaded.
Start bundle with: conduct run 73595ec
Unload bundle with: conduct unload 73595ec
Print ConductR info with: conduct info
```


### conduct load - private repository - No Bintray credentials

```
$ conduct load typesafe/internal-bundle/sherpa-akka
Retrieving bundle..
Loading bundle from cache typesafe/internal-bundle/sherpa-akka
Error: Unable to find Bintray repository typesafe/internal-bundle. If this is a private repository make sure to setup the Bintray credentials at /Users/mj/.lightbend/commercial.credentials
```

File: conductr_cli/sandbox_run_jvm.py [hunk 1]
```diff
@@ -269,6 +270,19 @@ def validate_64bit_support():
...
+def validate_bintray_credentials(image_version, offline_mode):
+    """
+    Validates if the necessary Bintray credentials exists to start the ConductR cluster.
+    Credentials are necessary prior to ConductR 2.1.0. From onwards 2.1.0, no credentials are needed.
+    :param image_version: the ConductR image version
+    :param offline_mode: the offline mode flag
+    """
+    if not offline_mode:
+        major_version, minor_version, _ = version_parts(image_version)
+        if major_version == 2 and minor_version == 0:
```
Comment (markusjura) [inline conductr_cli/sandbox_run_jvm.py:282]: The conductr version validation logic will be replaced by another PR. In there, we will put all ConductR version validation into a separate python file. This file will contain functions to query if certain capabilities based on the ConductR version exist. Influenced by https://en.wikipedia.org/wiki/Gestalt_(Mac_OS)

**Review by huntc — APPROVED** [review]
